### ISSUE TITLE
Update docs on KubernetesJobEnvironment to account for non-Docker storage

### ISSUE THREAD

**Issue by jameslamb** [PrefectHQ/prefect#2827]
## Current behavior

Using a flow registered with prefect cloud, with the following mix:

* `KubernetesJobEnvironment`
* `KubernetesAgent`
* `S3` storage
* `flow.register()` registering with Prefect Cloud

I see the following error in the agent's logs (I originally put this in https://github.com/PrefectHQ/prefect/issues/2791#issuecomment-646218984 but that might have been mistaken)

> [2020-06-18 17:55:21,643] ERROR - agent | Error while deploying flow: ValueError('Storage for flow run 1538c4c2-30a1-439a-b328-a1a40ec1e0dd is not of type Docker and environment has no image attribute in the metadata field.')

I think that the issue is here, in client code used by `flow.register()`

https://github.com/PrefectHQ/prefect/blob/b2de11b3e76d846c65d26a2c9ba9c6027a74cbfd/src/prefect/client/client.py#L651-L652

To me, it looks like the logic is

> "Agents that create jobs in docker containers need to know the image to use for the container. That image either needs to be set in the `environment`'s `["metadata"]["image"]`, or it will fall back to image used on the storage object if that is `Docker` storage"

This logic totally makes sense to me now that I see what's going on, but I couldn't form a good expectation about that from the documentation.

Once I changed my `KubernetesJobEnvironment` code to something like this, that error was resolved:

```python
env = KubernetesJobEnvironment(
    job_spec_file="prefect-flow-run.yaml",
    metadata={
        "image": "prefecthq/prefect:all_extras-0.12.0"
    }
)
```

Now that it's possible to use non-`Docker` storage with `KubernetesAgent` (#2666), I think the documentation at https://docs.prefect.io/orchestration/execution/k8s_job_environment.html#process could be improved.

## Proposed behavior

Today, the docs for `image:` at https://docs.prefect.io/orchestration/execution/k8s_job_environment.html#process  say

> image will become the registry_url/image_name:image_tag of your Flow's storage

I think it should be changed to something like

> image will become the registry_url/image_name:image_tag of your Flow's storage if you use Docker storage. If using non-docker storage, the image needs to be included in `KubernetesJobEnvironment`'s `metadata`. See https://docs.prefect.io/api/latest/environments/execution.html#kubernetesjobenvironment for details.

Similarly, I think the docs in the API Reference for `KubernetesJobEnvironment` (https://docs.prefect.io/api/latest/environments/execution.html#kubernetesjobenvironment) could be improved. Today, the documentation for `metadata` there says

> extra metadata to be set and serialized on this environment

I think it would be clearer to make that

> Extra metadata to be set and serialized on this environment. For example, if using non-Docker storage and a docker agent such as `KubernetesAgent`, you should add the desired image for flow runs here, such as `{"metadata": {"image": "prefecthq/prefect:0.12.0-python3.7}}`

At this point, I only know about `KubernetesAgent`. But my guess is that similar changes could be made for `FargateTaskEnvironment` and `DaskKubernetesEnvironment`.

## Example

I think these changes might help others avoid the debugging time I spent figuring out that the image needs to be set on `metadata` in the `KubernetesJobEnvironment` if using a similar mix of agent, storage, and environment.

If I've missed some place in the documentation where this is already described, I apologize.

Thanks for your time and consideration.

**Comment by jameslamb** [comment]
Oh! I forgot to add environment information. 

I'm using `prefect` 0.12.0, and my agent is running an image as of yesterday morning (after #2666 was merged), built like this:

```shell
# peg to a specific state of the Prefect repo
COMMIT_TO_USE="4ac1c8c14e0e85c470e552f5ebab3e07a5891ff4"
REGISTRY_URL="localhost:32000"
IMAGE_TAG="localhost:32000/prefect/prefect-test:${COMMIT_TO_USE:0:6}"

PREFECT_INSTALL_DIR=$(pwd)/prefect-temp
mkdir -p ${PREFECT_INSTALL_DIR}

pushd ${PREFECT_INSTALL_DIR}
    git clone git@github.com:PrefectHQ/prefect.git
    
    pushd prefect
        git checkout ${COMMIT_TO_USE}
        docker build \
            --build-arg PYTHON_VERSION=3.7 \
            --build-arg GIT_SHA=${COMMIT_TO_USE} \
            --build-arg BUILD_DATE="today" \
            --build-arg PREFECT_VERSION=${COMMIT_TO_USE} \
            --tag ${IMAGE_TAG} \
            -f Dockerfile \
            .
        docker push ${IMAGE_TAG}
    popd

popd
```

**Comment by joshmeek** [comment]
Ah @jameslamb yes the documentation on the environments pages needs to be updated thanks for letting me know! Currently it's slightly hidden here https://docs.prefect.io/orchestration/execution/storage_options.html#non-docker-storage-for-containerized-environments

**Comment by jameslamb** [comment]
ah! that's exactly what I was looking for!

I even touched that section of the docs a few days ago (#2517 ), I should have recognized that :joy: 

I think it didn't click for me because my thought process while working through this was "I want to use a `KubernetesJobEnvironment`, let me go look at those docs", instead of "I'm using `S3` storage, how does that affect the way I set up the environment"

**Comment by jcrist** [comment]
With the 0.14.0 release, run-configs have replaced environments as the way to configure flow runs. Any flow can provide an image by passing in the `image` kwarg to its corresponding run-config. For example, if deploying on k8s, you'd do:

```python
from prefect.run_configs import KubernetesRun

flow.run_config = KubernetesRun(image="my-fancy-image")
```

This should (hopefully) be well documented in the [flow configuration](https://docs.prefect.io/orchestration/flow_config/run_configs.html) docs. Closing for now, feel free to reopen (or open a new issue) if you find the docs still lacking in this regard.
